pyspark_datasources/googlesheets.py

```diff
@@ -0,0 +1,160 @@
+    |country|  latitude|  longitude|                name|
+    +-------+----------+-----------+--------------------+
+    |     AD| 42.546245|   1.601554|             Andorra|
+    |    ...|       ...|        ...|                 ...|
+    +-------+----------+-----------+--------------------+
+
+    Load data using `spreadsheet_id` and optional `sheet_id`.
+
+    >>> spark.read.format("googlesheets").options(spreadsheet_id="10pD8oRN3RTBJq976RKPWHuxYy0Qa_JOoGFpsaS0Lop0", sheet_id="0")
+    +-------+----------+-----------+--------------------+
+    |country|  latitude|  longitude|                name|
+    +-------+----------+-----------+--------------------+
+    |     AD| 42.546245|   1.601554|             Andorra|
+    |    ...|       ...|        ...|                 ...|
+    +-------+----------+-----------+--------------------+
+
+    Specify custom schema.
+
+    >>> spark.read.format("googlesheets").options(url=url).schema("id string, lat double, long double, name string")
+    +---+----------+-----------+--------------------+
+    | id|       lat|       long|                name|
+    +---+----------+-----------+--------------------+
+    | AD| 42.546245|   1.601554|             Andorra|
+    |...|       ...|        ...|                 ...|
+    +---+----------+-----------+--------------------+
+    """
+
+    @classmethod
+    def name(self):
+        return "googlesheets"
+
+    def __init__(self, options: Dict[str, str]):
+        if "url" in options:
+            self.sheet = Sheet.from_url(options["url"])
+        elif "spreadsheet_id" in options:
+            self.sheet = Sheet(options["spreadsheet_id"], options.get("sheet_id"))
+        else:
+            raise ValueError(
+                "You must specify a URL or spreadsheet_id in `.options()`."
+            )
+
+    def schema(self) -> StructType:
+        import pandas as pd
+
+        # Read schema from the first row of the sheet
+        df = pd.read_csv(self.sheet.get_query_url("select * limit 0"))
+        return StructType([StructField(col, StringType()) for col in df.columns])
+
+    def reader(self, schema: StructType) -> DataSourceReader:
+        return GoogleSheetsReader(self.sheet, schema)
+
+
+class GoogleSheetsReader(DataSourceReader):
+    def __init__(self, sheet: Sheet, schema: StructType):
+        self.sheet = sheet
+        self.schema = schema
+
+    def read(self, partition):
+        from urllib.request import urlopen
+
+        from pyarrow import csv
+        from pyspark.sql.pandas.types import to_arrow_type
+
+        # Specify column types based on the schema
+        convert_options = csv.ConvertOptions(
+            column_types={
+                field.name: to_arrow_type(field.dataType) for field in self.schema
+            },
+        )
+        read_options = csv.ReadOptions(
+            column_names=self.schema.fieldNames(),  # Rename columns
+            skip_rows=1,  # Skip the header row
```

Comment:
What is the spreadsheet does not have a header row?

Reply:
We can add an option for `skip_rows`